### PR TITLE
feat(speechchallenge): add srt and image file parameters

### LINKED ISSUE #229: Enhance SpeechChallange
The recording challenge needs additional fields that the backend will provide soon

- SRT file
- Image

### PR DESCRIPTION
**Pull request by Lehren**
Add srt and image file parameters to the speechchallenge constructor.
Add srt and image file options when creating a speechchallenge.

Fixes #229

### CODE REVIEW

File: src/administrative-sdk/speech-challenge/speech-challenge-controller.js [hunk 1]
```diff
@@ -18,20 +18,26 @@ export default class SpeechChallengeController {
...
    *
    * @param {SpeechChallenge} speechChallenge - Object to create.
    * @param {?Blob} audioBlob - Audio fragment to link to the challenge.
+   * @param {?Blob} srtFile - SRT file in HTML5 Blob format to accompany the challenge.
+   * @param {?Blob} image - Image file in HTML5 Blob format to accompany the challenge.
    * @returns {Promise.<PronunciationChallenge>} Promise containing the newly created SpeechChallenge.
    * @throws {Promise.<Error>} If the server returned an error.
    */
-  createSpeechChallenge(speechChallenge, audioBlob) {
+  createSpeechChallenge(speechChallenge, audioBlob, srtFile, image) {
```
Comment (joggienl) [inline src/administrative-sdk/speech-challenge/speech-challenge-controller.js]: I think we could or should be using object destructoring more. It saves long argument lists and makes it more easy to write pure functions.

File: src/administrative-sdk/speech-challenge/speech-challenge-controller.js [hunk 2]
```diff
@@ -54,7 +60,7 @@ export default class SpeechChallengeController {
 
     return this._connection._secureAjaxGet(url)
       .then(data => {
-        const challenge = new SpeechChallenge(data.id, data.topic, data.referenceAudioUrl);
+        const challenge = new SpeechChallenge(data.id, data.topic, data.referenceAudioUrl, data.srtUrl, data.imageUrl);
```
Comment (joggienl) [inline src/administrative-sdk/speech-challenge/speech-challenge-controller.js]: If you would apply object destructoring you would only have to pass the `data` object to the SpeechChallenge constructor.

File: src/administrative-sdk/speech-challenge/speech-challenge.js [hunk 3]
```diff
@@ -9,8 +9,12 @@ export default class SpeechChallenge {
    * @param {?string} topic - A question or topic serving as guidance.
    * @param {?string} referenceAudioUrl - The reference audio fragment URL. If one is not yet available or audio is
    * not yet registered to the challenge it can be set to 'null'.
+   * @param {?string} srtUrl - URL of a possible .srt file to accompany this challenge.
+   * @param {?string} imageUrl - URL of a possible image file to accompany this challenge.
+   * @throws {Error} srtUrl parameter of type "string|null" is required.
+   * @throws {Error} imageUrl parameter of type "string|null" is required.
    */
-  constructor(id, topic, referenceAudioUrl) {
+  constructor(id, topic, referenceAudioUrl, srtUrl = null, imageUrl = null) {
```
Comment (joggienl) [inline src/administrative-sdk/speech-challenge/speech-challenge.js]: If you would want to apply object destructoring, just change this line to:

```js
constructor({id, topic, referenceAudioUrl, srtUrl = null, imageUrl = null}) {
  // id, topic, referenceAudioUrl, srtUrl, imageUrl is available.
}
```

Why not also accept the creation/update fields here?

File: src/administrative-sdk/speech-challenge/speech-challenge-controller.js [hunk 4]
```diff
@@ -18,20 +18,26 @@ export default class SpeechChallengeController {
...
+    speechChallenge.srt = srtFile;
+    speechChallenge.image = image;
     const fd = JSON.stringify(speechChallenge);
     const url = this._connection._settings.apiUrl + '/challenges/speech';
 
     return this._connection._secureAjaxPost(url, fd)
       .then(data => {
-        const result = new SpeechChallenge(data.id, data.topic, data.referenceAudioUrl);
+        const result = new SpeechChallenge(data.id, data.topic, data.referenceAudioUrl, data.srtUrl, data.imageUrl);
         result.created = new Date(data.created);
```
Comment (Lehren) [inline src/administrative-sdk/speech-challenge/speech-challenge-controller.js]: User made objects should never have the updated and created attributes. Those only come from the API.

File: src/administrative-sdk/speech-challenge/speech-challenge-controller.js (hunk 1 above)
Comment (Lehren) [inline src/administrative-sdk/speech-challenge/speech-challenge-controller.js]: Oh that's a new thing for me. Looks useful!

File: src/administrative-sdk/speech-challenge/speech-challenge-controller.js (hunk 4 above)
Comment (joggienl) [inline src/administrative-sdk/speech-challenge/speech-challenge-controller.js]: Yes, but you could pass them to the object for initialisation

File: src/administrative-sdk/speech-challenge/speech-challenge.js (hunk 3 above)
Comment (Lehren) [inline src/administrative-sdk/speech-challenge/speech-challenge.js]: What happens to regular object construction then?

`const org = new Organisation('identifier', 'orgname');`

turns into?

`const org = new Organisation({id: 'identifier', name: 'orgname'});`

Is constructor overloading a thing? Don't think so...

Comment (joggienl) [inline src/administrative-sdk/speech-challenge/speech-challenge.js]: If it takes one or two parameters it doesn't make a lot of sense.

You'll have to pass in an object indeed. But most often you already have an object from which you're creating objects from.

**Comment by joggienl** [comment]
Had some offline discussion with @Lehren about destructoring (and this concept in more general). Possible PR's with this could be coming. Accepting this for now as is!

**Pull request merged**